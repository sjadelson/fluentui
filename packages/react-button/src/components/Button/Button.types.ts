import * as React from 'react';
import { ComponentProps, ShorthandProps } from '@fluentui/react-compose/lib/next/index';
import { ColorTokens, RecursivePartial, SizeValue, FontTokens } from '@fluentui/theme';

/**
 * {@docCategory Button}
 */
export type ButtonProps = ComponentProps &
  React.HTMLAttributes<HTMLElement> & {
    /**
     * Shorthand icon. A shorthand prop can be a literal, object, or
     * JSX. The `children` prop of the object can be a render function,
     * taking in the original slot component and props.
     */
    icon?: ShorthandProps;

    /**
     * Shorthand loader content within the button.
     */
    loader?: ShorthandProps;

    /**
     * Shorthand children content within the button.
     */
    content?: ShorthandProps;

    /**
     * Defines the href to navigate to. If applied, will render the button as an anchor
     * element by default, unless `as` specifies otherwise. Note that specifying an href
     * and a non-anchor as the render type will prevent the Button from behaving like a
     * hyperlink and opening the href on click.
     */
    href?: string;

    /**
     * Defines the target window to open the href in. Only is applied if the button renders
     * as an anchor tag, which is the default behavior if href is provided.
     */
    target?: string;

    /** A button can appear circular. */
    circular?: boolean;

    /** A button can show that it cannot be interacted with. */
    disabled?: boolean;

    /**
     * When set, allows the button to be focusable even when it has been disabled. This is used in scenarios where it is
     * important to keep a consistent tab order for screen reader and keyboard users.
     * @defaultvalue false
     */
    disabledFocusable?: boolean;

    /** A button can fill the width of its container. */
    block?: boolean;

    /** A button can contain only an icon. */
    iconOnly?: boolean;

    /** An icon button can format its icon to appear before or after its content. */
    iconPosition?: 'before' | 'after';

    /** A button that inherits its background and has a subtle appearance. */
    inverted?: boolean;

    /** Opt in to a named variant which can be supplied by a theme. */
    variant?: string;

    /** A button can show a loading indicator. */
    loading?: boolean;

    /**
     * Called after a user clicks the button.
     * @param event - React's original SyntheticEvent.
     * @param data - All props.
     */
    // onClick?: ComponentEventHandler<ButtonProps>;

    /**
     * Called after a user focuses the button.
     * @param event - React's original SyntheticEvent.
     * @param data - All props.
     */
    // onFocus?: ComponentEventHandler<ButtonProps>;

    /** A button can emphasize that it represents the primary action. */
    primary?: boolean;

    /** A button can emphasize that it represents an alternative action. */
    secondary?: boolean;

    /** A button can blend into its background to become less emphasized. */
    ghost?: boolean;

    /** A button can have no background styling and just be emphasized through its content styling. */
    transparent?: boolean;

    /** A button can be sized. */
    size?: SizeValue;

    // TODO: Deprecate or rename to textOnly for alignment with iconOnly?
    /** A button can be formatted to show only text in order to indicate a less-pronounced action. */
    // text?: boolean;

    tokens?: RecursivePartial<ButtonTokens>;
  };

/**
 * {@docCategory Button}
 */
export interface ButtonState extends ButtonProps {
  buttonRef?: React.RefObject<HTMLButtonElement>;
}

/**
 * {@docCategory Button}
 */
export type ButtonTokens = ColorTokens &
  FontTokens & {
    paddingLeft?: string;
    paddingRight?: string;
    paddingTop?: string;
    paddingBottom?: string;
    margin?: string;
    height?: string;
    minWidth?: string;
    maxWidth?: string;
    minHeight?: string;
    contentGap?: string;
    iconSize?: string;
    borderRadius?: string;
    borderTopLeftRadius?: string;
    borderTopRightRadius?: string;
    borderBottomLeftRadius?: string;
    borderBottomRightRadius?: string;
    borderWidth?: string;
    boxShadow?: string;
    width?: string;

    forcedColorAdjust?: string;

    transform?: string;
    transition?: string;

    /* sizing */
    size?: {
      smallest?: string;
      smaller?: string;
      small?: string;
      regular?: string;
      large?: string;
      larger?: string;
      largest?: string;
    };

    pressed?: {
      transform?: string;
      transition?: string;
    };
  };

/**
 * {@docCategory Button}
 */
export type ButtonVariants<TTokens = ButtonTokens> = {
  root?: TTokens;
  block?: TTokens;
  circular?: TTokens;
  iconOnly?: TTokens;
  ghost?: TTokens;
  primary?: TTokens;
  transparent?: TTokens;
};
