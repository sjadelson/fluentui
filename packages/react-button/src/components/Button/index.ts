export * from './Button.types';
export * from './Button';
export * from './renderButton';
export * from './useButton';
export * from './useButtonClasses';
export * from './useButtonState';
