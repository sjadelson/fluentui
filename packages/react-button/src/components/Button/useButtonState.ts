import * as React from 'react';
import { getCode, EnterKey, SpacebarKey } from '@fluentui/keyboard-key';
import { ButtonState } from './Button.types';

/**
 * The useButton hook processes the Button draft state.
 * @param draftState - Button draft state to mutate.
 */
export const useButtonState = (draftState: ButtonState) => {
  // Update the button's tab-index, keyboard handling, and aria attributes.
  if (draftState.as !== 'button') {
    draftState.role = 'button';

    if (draftState.as !== 'a') {
      const { onClick: onClickCallback, onKeyDown: onKeyDownCallback } = draftState;

      draftState['data-isFocusable'] = true;
      draftState.tabIndex = 0;

      draftState.onKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
        if (onKeyDownCallback) {
          onKeyDownCallback(ev);
        }

        const keyCode = getCode(ev);
        if (!ev.defaultPrevented && onClickCallback && (keyCode === EnterKey || keyCode === SpacebarKey)) {
          // Translate the keydown enter/space to a click.
          ev.preventDefault();
          ev.stopPropagation();

          (ev.target as HTMLElement).click();
        }
      };
    }
  }

  // Disallow click and keyboard events when component is disabled and eat events when disabledFocusable is set to true.
  const { disabled, onKeyDown } = draftState;
  if (disabled) {
    draftState.onClick = undefined;
  }
  draftState.onKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
    const keyCode = getCode(ev);
    if (disabled && (keyCode === EnterKey || keyCode === SpacebarKey)) {
      ev.preventDefault();
      ev.stopPropagation();
    } else {
      onKeyDown?.(ev);
    }
  };

  draftState['aria-disabled'] = draftState.disabled;
  draftState.disabled =
    draftState.as === 'button' ? draftState['aria-disabled'] && !draftState.disabledFocusable : undefined;
};
