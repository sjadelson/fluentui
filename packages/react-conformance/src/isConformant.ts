import * as fs from 'fs';

import { IsConformantOptions } from './types';
import { defaultTests } from './defaultTests';
import { defaultErrorMessages } from './defaultErrorMessages';
import { merge } from './utils/merge';
import os from 'os';
import chalk from 'chalk';
import { getComponentDoc } from './utils/getComponentDoc';

export function isConformant(...testInfo: Partial<IsConformantOptions>[]) {
  const mergedOptions = merge<IsConformantOptions>(...testInfo);
  const { componentPath, displayName, disabledTests = [], extraTests, isInternal, exportSubdir = '' } = mergedOptions;
  if (!fs.existsSync(componentPath)) {
    throw new Error(`Path ${componentPath} does not exist`);
  }

  const components = getComponentDoc(componentPath, exportSubdir);
  const mainComponents = components.filter(comp => comp.displayName === displayName);

  if (mainComponents.length === 1) {
    const componentInfo = mainComponents[0];

    if (isInternal) {
      disabledTests.push('exported-top-level');
    }

    describe('isConformant', () => {
      afterAll(() => {
        defaultErrorMessages['display-failed-tests'](componentInfo, mergedOptions);
      });

      for (const test of Object.keys(defaultTests)) {
        if (!disabledTests.includes(test)) {
          defaultTests[test](componentInfo, mergedOptions);
        }
      }
    });

    if (extraTests) {
      describe('isConformant - extraTests', () => {
        for (const test of Object.keys(extraTests)) {
          extraTests[test](componentInfo, mergedOptions);
        }
      });
    }
  } else if (components.length === 0) {
    console.log(
      chalk.yellow(`No exported components in path: `) + os.EOL.repeat(2) + chalk.green.italic(componentPath),
    );
    throw new Error('No exported components in path');
  } else {
    console.log(
      chalk.yellow(`No component with name `) +
        chalk.hex('#e00000')(displayName) +
        chalk.yellow(' was found at:') +
        os.EOL.repeat(2) +
        chalk.green.italic(componentPath) +
        os.EOL.repeat(2) +
        'These are the exported component names:' +
        os.EOL.repeat(2) +
        chalk.white.bold.italic.bgHex('#2e2e2e')(components.map(component => component.displayName).join(', ')),
    );
    throw new Error(`No component was found that matches the displayName.`);
  }
}
