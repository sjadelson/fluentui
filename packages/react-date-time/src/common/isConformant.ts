import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'> & { componentPath?: string }) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes'],
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
