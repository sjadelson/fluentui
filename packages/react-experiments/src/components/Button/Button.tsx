import * as React from 'react';
// Temporary import file to experiment with next version of foundation.
import { composed } from '@fluentui/foundation-legacy/lib/next/composed';
import { useButtonState as state } from './Button.state';
import { ButtonStyles as styles, ButtonTokens as tokens } from './Button.styles';
import { IButtonProps } from './Button.types';
import { ButtonSlots as slots, ButtonView as view } from './Button.view';

export const Button: React.FunctionComponent<IButtonProps> = composed({
  displayName: 'Button',
  slots,
  state,
  styles,
  tokens,
  view,
});

export default Button;
