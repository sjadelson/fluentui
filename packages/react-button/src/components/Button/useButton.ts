import * as React from 'react';
import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-compose/lib/next/index';
import { ButtonProps, ButtonState } from './Button.types';
import { useButtonState } from './useButtonState';

/**
 * Consts listing which props are shorthand props.
 */
export const buttonShorthandProps = ['icon', 'loader', 'content'];

const mergeProps = makeMergeProps({ deepMerge: buttonShorthandProps });

/**
 * Given user props, returns state and render function for a Button.
 */
export const useButton = (props: ButtonProps, ref: React.Ref<HTMLElement>, defaultProps?: ButtonProps) => {
  // Ensure that the `ref` prop can be used by other things (like useFocusRects) to refer to the root.
  // NOTE: We are assuming refs should not mutate to undefined. Either they are passed or not.
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const resolvedRef = ref || React.useRef();
  const state = mergeProps(
    {
      ref: resolvedRef,
      as: props.href ? 'a' : 'button',
      icon: { as: 'span' },
      content: { as: 'span', children: props.children },
      loader: { as: 'span' },
    },
    defaultProps,
    resolveShorthandProps(props, buttonShorthandProps),
  );

  useButtonState(state);

  return state as ButtonState;
};
