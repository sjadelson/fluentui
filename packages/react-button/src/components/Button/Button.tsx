import * as React from 'react';
import { useButton } from './useButton';
import { ButtonProps } from './Button.types';
import { useInlineTokens } from '@fluentui/react-theme-provider';
import { useButtonClasses } from './useButtonClasses';

/**
 * Define a styled Button, using the `useButton` hook.
 * {@docCategory Button}
 */
export const Button = React.forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const { render, state } = useButton(props, ref);

  useButtonClasses(state);
  useInlineTokens(state, '--button');

  return render(state);
});

Button.displayName = 'Button';
