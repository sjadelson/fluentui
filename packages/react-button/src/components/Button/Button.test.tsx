import * as React from 'react';
import { Button } from './Button';
import * as renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('Button (isConformant)', () =>
  isConformant({
    Component: Button,
    displayName: 'Button',
  }));

describe('Button', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Button in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Button>Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders anchor when href prop is provided', () => {
    const component = renderer.create(<Button href="https://www.bing.com">Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can be focused', () => {
    const rootRef = React.createRef<HTMLButtonElement>();

    wrapper = mount(<Button ref={rootRef}>Focus me</Button>);

    expect(typeof rootRef.current).toEqual('object');
    expect(document.activeElement).not.toEqual(rootRef.current);

    rootRef.current?.focus();

    expect(document.activeElement).toEqual(rootRef.current);
  });
});
