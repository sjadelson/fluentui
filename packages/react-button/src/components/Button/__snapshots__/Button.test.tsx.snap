// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a default state 1`] = `
<button
  className=
      ms-Button
      {
        --button-background: #ffffff;
        --button-borderColor: #8a8886;
        --button-borderRadius: 2px;
        --button-borderWidth: 1px;
        --button-boxShadow: none;
        --button-contentColor: #323130;
        --button-contentGap: 8px;
        --button-disabled-background: #f3f2f1;
        --button-disabled-borderColor: #f3f2f1;
        --button-disabled-contentColor: #a19f9d;
        --button-focusColor: #000000;
        --button-focusInnerColor: #ffffff;
        --button-fontFamily: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        --button-fontSize: 14px;
        --button-fontWeight: 600;
        --button-highContrast-borderColor: WindowText;
        --button-highContrast-contentColor: WindowText;
        --button-highContrast-disabled-background: Window;
        --button-highContrast-disabled-borderColor: GrayText;
        --button-highContrast-disabled-contentColor: GrayText;
        --button-highContrast-disabled-iconColor: GrayText;
        --button-highContrast-hovered-borderColor: Highlight;
        --button-highContrast-hovered-contentColor: Highlight;
        --button-highContrast-hovered-iconColor: Highlight;
        --button-highContrast-iconColor: WindowText;
        --button-highContrast-pressed-borderColor: Highlight;
        --button-highContrast-pressed-contentColor: WindowText;
        --button-highContrast-pressed-iconColor: WindowText;
        --button-hovered-background: #f3f2f1;
        --button-hovered-contentColor: #201f1e;
        --button-iconColor: inherit;
        --button-iconSize: 16px;
        --button-margin: 0;
        --button-maxWidth: 280px;
        --button-minHeight: var(--button-size-regular);
        --button-minWidth: 96px;
        --button-paddingBottom: 0;
        --button-paddingLeft: 20px;
        --button-paddingRight: 20px;
        --button-paddingTop: 0;
        --button-pressed-background: #edebe9;
        --button-pressed-contentColor: #201f1e;
        --button-size-large: 40px;
        --button-size-larger: 48px;
        --button-size-largest: 64px;
        --button-size-regular: 32px;
        --button-size-small: 24px;
        --button-size-smaller: 24px;
        --button-size-smallest: 24px;
        --button-width: auto;
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        align-items: center;
        background: var(--button-background);
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-bottom-width: var(--button-borderBottomWidth, var(--button-borderWidth));
        border-color: var(--button-borderColor);
        border-left-width: var(--button-borderLeftWidth, var(--button-borderWidth));
        border-right-width: var(--button-borderRightWidth, var(--button-borderWidth));
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-top-width: var(--button-borderTopWidth, var(--button-borderWidth));
        box-shadow: var(--button-boxShadow);
        box-sizing: border-box;
        color: var(--button-contentColor);
        cursor: pointer;
        display: inline-flex;
        font-family: var(--button-fontFamily);
        font-size: var(--button-fontSize);
        font-weight: var(--button-fontWeight);
        height: var(--button-height);
        justify-content: center;
        max-width: var(--button-maxWidth);
        min-height: var(--button-minHeight);
        min-width: var(--button-minWidth);
        outline: none;
        padding-bottom: var(--button-paddingBottom);
        padding-left: var(--button-paddingLeft);
        padding-right: var(--button-paddingRight);
        padding-top: var(--button-paddingTop);
        position: relative;
        text-decoration: none;
        transition: var(--button-transition);
        user-select: none;
        vertical-align: middle;
        white-space: var(--button-whiteSpace);
        width: var(--button-width);
      }
      .ms-Fabric--isFocusVisible &:focus::after {
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-color: var(--button-focusColor, black);
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-width: var(--button-focusWidth, 2px);
        bottom: -1px;
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-focusInnerColor, white) inset;
        content: "";
        left: -1px;
        position: absolute;
        right: -1px;
        top: -1px;
        z-index: 1;
      }
      @media screen and (forced-colors: active){.ms-Fabric--isFocusVisible &:focus::after {
        border-color: var(--button-highContrast-focusColor);
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-highContrast-focusInnerColor) inset;
      }
      & > *:not(:first-child) {
        margin-left: var(--button-contentGap);
      }
      @media screen and (forced-colors: active){& {
        background: var(--button-highContrast-background);
        border-color: var(--button-highContrast-borderColor);
        color: var(--button-highContrast-contentColor);
        forced-color-adjust: var(--button-forcedColorAdjust);
      }
      @media screen and (forced-colors: active){& .ms-Button-icon {
        color: var(--button-highContrast-iconColor);
      }
      &:hover {
        background: var(--button-hovered-background, var(--button-background));
        border-color: var(--button-hovered-borderColor, var(--button-borderColor));
        box-shadow: var(--button-hovered-boxShadow, var(--button-boxShadow));
        color: var(--button-hovered-contentColor, var(--button-contentColor));
      }
      &:hover .ms-Button-icon {
        color: var(--button-hovered-iconColor, var(--button-iconColor));
      }
      @media screen and (forced-colors: active){&:hover {
        background: var(--button-highContrast-hovered-background, var(--button-highContrast-background));
        border-color: var(--button-highContrast-hovered-borderColor, var(--button-highContrast-borderColor));
        color: var(--button-highContrast-hovered-contentColor, var(--button-highContrast-contentColor));
      }
      @media screen and (forced-colors: active){&:hover .ms-Button-icon {
        color: var(--button-highContrast-hovered-iconColor, --button-highContrast-iconColor);
      }
      &:active {
        background: var(--button-pressed-background, var(--button-hovered-background));
        border-color: var(--button-pressed-borderColor, var(--button-hovered-borderColor, var(--button-borderColor)));
        box-shadow: var(--button-pressed-boxShadow, var(--button-hovered-boxShadow, var(--button-boxShadow)));
        color: var(--button-pressed-contentColor, var(--button-hovered-contentColor, var(--button-contentColor)));
        transform: var(--button-pressed-transform);
        transition: var(--button-pressed-transition);
      }
      &:active .ms-Button-icon {
        color: var(--button-pressed-iconColor, var(--button-iconColor));
      }
      @media screen and (forced-colors: active){&:active {
        background: var(--button-highContrast-pressed-background, var(--button-highContrast-hovered-background, var(--button-highContrast-background)));
        border-color: var(--button-highContrast-pressed-borderColor, var(--button-highContrast-hovered-borderColor, var(--button-highContrast-borderColor)));
        color: var(--button-highContrast-pressed-contentColor, var(--button-highContrast-hovered-contentColor, var(--button-highContarst-contentColor)));
      }
      @media screen and (forced-colors: active){&:active .ms-Button-icon {
        color: var(--button-highContrast-pressed-iconColor, var(--button-highContrast-hovered-iconColor, var(--button-highContrast-iconColor)));
      }
      &[aria-disabled=true] {
        background-color: var(--button-disabled-background);
        border-color: var(--button-disabled-borderColor);
        box-shadow: var(--button-disabled-boxShadow);
        color: var(--button-disabled-contentColor);
        opacity: var(--button-disabled-opacity);
        pointer-events: none;
      }
      &[aria-disabled=true] .ms-Button-icon {
        color: var(--button-disabled-iconColor);
      }
      @media screen and (forced-colors: active){&[aria-disabled=true] {
        background: var(--button-highContrast-disabled-background, var(--button-highContrast-background));
        border-color: var(--button-highContrast-disabled-borderColor, var(--button-highContrast-borderColor));
        color: var(--button-highContrast-disabled-contentColor, var(--button-highContrast-contentColor));
      }
      @media screen and (forced-colors: active){&[aria-disabled=true] .ms-Button-icon {
        color: var(--button-highContrast-disabled-iconColor, --button-highContrast-iconColor);
      }
>
  <span
    className=

        {
          line-height: 1;
        }
  >
    Default button
  </span>
</button>
`;

exports[`Button renders anchor when href prop is provided 1`] = `
<a
  className=
      ms-Button
      {
        --button-background: #ffffff;
        --button-borderColor: #8a8886;
        --button-borderRadius: 2px;
        --button-borderWidth: 1px;
        --button-boxShadow: none;
        --button-contentColor: #323130;
        --button-contentGap: 8px;
        --button-disabled-background: #f3f2f1;
        --button-disabled-borderColor: #f3f2f1;
        --button-disabled-contentColor: #a19f9d;
        --button-focusColor: #000000;
        --button-focusInnerColor: #ffffff;
        --button-fontFamily: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        --button-fontSize: 14px;
        --button-fontWeight: 600;
        --button-highContrast-borderColor: WindowText;
        --button-highContrast-contentColor: WindowText;
        --button-highContrast-disabled-background: Window;
        --button-highContrast-disabled-borderColor: GrayText;
        --button-highContrast-disabled-contentColor: GrayText;
        --button-highContrast-disabled-iconColor: GrayText;
        --button-highContrast-hovered-borderColor: Highlight;
        --button-highContrast-hovered-contentColor: Highlight;
        --button-highContrast-hovered-iconColor: Highlight;
        --button-highContrast-iconColor: WindowText;
        --button-highContrast-pressed-borderColor: Highlight;
        --button-highContrast-pressed-contentColor: WindowText;
        --button-highContrast-pressed-iconColor: WindowText;
        --button-hovered-background: #f3f2f1;
        --button-hovered-contentColor: #201f1e;
        --button-iconColor: inherit;
        --button-iconSize: 16px;
        --button-margin: 0;
        --button-maxWidth: 280px;
        --button-minHeight: var(--button-size-regular);
        --button-minWidth: 96px;
        --button-paddingBottom: 0;
        --button-paddingLeft: 20px;
        --button-paddingRight: 20px;
        --button-paddingTop: 0;
        --button-pressed-background: #edebe9;
        --button-pressed-contentColor: #201f1e;
        --button-size-large: 40px;
        --button-size-larger: 48px;
        --button-size-largest: 64px;
        --button-size-regular: 32px;
        --button-size-small: 24px;
        --button-size-smaller: 24px;
        --button-size-smallest: 24px;
        --button-width: auto;
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        align-items: center;
        background: var(--button-background);
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-bottom-width: var(--button-borderBottomWidth, var(--button-borderWidth));
        border-color: var(--button-borderColor);
        border-left-width: var(--button-borderLeftWidth, var(--button-borderWidth));
        border-right-width: var(--button-borderRightWidth, var(--button-borderWidth));
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-top-width: var(--button-borderTopWidth, var(--button-borderWidth));
        box-shadow: var(--button-boxShadow);
        box-sizing: border-box;
        color: var(--button-contentColor);
        cursor: pointer;
        display: inline-flex;
        font-family: var(--button-fontFamily);
        font-size: var(--button-fontSize);
        font-weight: var(--button-fontWeight);
        height: var(--button-height);
        justify-content: center;
        max-width: var(--button-maxWidth);
        min-height: var(--button-minHeight);
        min-width: var(--button-minWidth);
        outline: none;
        padding-bottom: var(--button-paddingBottom);
        padding-left: var(--button-paddingLeft);
        padding-right: var(--button-paddingRight);
        padding-top: var(--button-paddingTop);
        position: relative;
        text-decoration: none;
        transition: var(--button-transition);
        user-select: none;
        vertical-align: middle;
        white-space: var(--button-whiteSpace);
        width: var(--button-width);
      }
      .ms-Fabric--isFocusVisible &:focus::after {
        border-bottom-left-radius: var(--button-borderBottomLeftRadius, var(--button-borderRadius));
        border-bottom-right-radius: var(--button-borderBottomRightRadius, var(--button-borderRadius));
        border-color: var(--button-focusColor, black);
        border-style: solid;
        border-top-left-radius: var(--button-borderTopLeftRadius, var(--button-borderRadius));
        border-top-right-radius: var(--button-borderTopRightRadius, var(--button-borderRadius));
        border-width: var(--button-focusWidth, 2px);
        bottom: -1px;
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-focusInnerColor, white) inset;
        content: "";
        left: -1px;
        position: absolute;
        right: -1px;
        top: -1px;
        z-index: 1;
      }
      @media screen and (forced-colors: active){.ms-Fabric--isFocusVisible &:focus::after {
        border-color: var(--button-highContrast-focusColor);
        box-shadow: 0 0 0 var(--button-focusInnerWidth, 1px) var(--button-highContrast-focusInnerColor) inset;
      }
      & > *:not(:first-child) {
        margin-left: var(--button-contentGap);
      }
      @media screen and (forced-colors: active){& {
        background: var(--button-highContrast-background);
        border-color: var(--button-highContrast-borderColor);
        color: var(--button-highContrast-contentColor);
        forced-color-adjust: var(--button-forcedColorAdjust);
      }
      @media screen and (forced-colors: active){& .ms-Button-icon {
        color: var(--button-highContrast-iconColor);
      }
      &:hover {
        background: var(--button-hovered-background, var(--button-background));
        border-color: var(--button-hovered-borderColor, var(--button-borderColor));
        box-shadow: var(--button-hovered-boxShadow, var(--button-boxShadow));
        color: var(--button-hovered-contentColor, var(--button-contentColor));
      }
      &:hover .ms-Button-icon {
        color: var(--button-hovered-iconColor, var(--button-iconColor));
      }
      @media screen and (forced-colors: active){&:hover {
        background: var(--button-highContrast-hovered-background, var(--button-highContrast-background));
        border-color: var(--button-highContrast-hovered-borderColor, var(--button-highContrast-borderColor));
        color: var(--button-highContrast-hovered-contentColor, var(--button-highContrast-contentColor));
      }
      @media screen and (forced-colors: active){&:hover .ms-Button-icon {
        color: var(--button-highContrast-hovered-iconColor, --button-highContrast-iconColor);
      }
      &:active {
        background: var(--button-pressed-background, var(--button-hovered-background));
        border-color: var(--button-pressed-borderColor, var(--button-hovered-borderColor, var(--button-borderColor)));
        box-shadow: var(--button-pressed-boxShadow, var(--button-hovered-boxShadow, var(--button-boxShadow)));
        color: var(--button-pressed-contentColor, var(--button-hovered-contentColor, var(--button-contentColor)));
        transform: var(--button-pressed-transform);
        transition: var(--button-pressed-transition);
      }
      &:active .ms-Button-icon {
        color: var(--button-pressed-iconColor, var(--button-iconColor));
      }
      @media screen and (forced-colors: active){&:active {
        background: var(--button-highContrast-pressed-background, var(--button-highContrast-hovered-background, var(--button-highContrast-background)));
        border-color: var(--button-highContrast-pressed-borderColor, var(--button-highContrast-hovered-borderColor, var(--button-highContrast-borderColor)));
        color: var(--button-highContrast-pressed-contentColor, var(--button-highContrast-hovered-contentColor, var(--button-highContarst-contentColor)));
      }
      @media screen and (forced-colors: active){&:active .ms-Button-icon {
        color: var(--button-highContrast-pressed-iconColor, var(--button-highContrast-hovered-iconColor, var(--button-highContrast-iconColor)));
      }
      &[aria-disabled=true] {
        background-color: var(--button-disabled-background);
        border-color: var(--button-disabled-borderColor);
        box-shadow: var(--button-disabled-boxShadow);
        color: var(--button-disabled-contentColor);
        opacity: var(--button-disabled-opacity);
        pointer-events: none;
      }
      &[aria-disabled=true] .ms-Button-icon {
        color: var(--button-disabled-iconColor);
      }
      @media screen and (forced-colors: active){&[aria-disabled=true] {
        background: var(--button-highContrast-disabled-background, var(--button-highContrast-background));
        border-color: var(--button-highContrast-disabled-borderColor, var(--button-highContrast-borderColor));
        color: var(--button-highContrast-disabled-contentColor, var(--button-highContrast-contentColor));
      }
      @media screen and (forced-colors: active){&[aria-disabled=true] .ms-Button-icon {
        color: var(--button-highContrast-disabled-iconColor, --button-highContrast-iconColor);
      }
  href="https://www.bing.com"
  role="button"
>
  <span
    className=

        {
          line-height: 1;
        }
  >
    Default button
  </span>
</a>
`;
